src/data/finsupp/to_dfinsupp.lean

```diff
@@ -0,0 +1,190 @@
+
+* `finsupp.to_dfinsupp_coe`
+* `finsupp.to_dfinsupp_support`
+* `dfinsupp.to_finsupp_coe`
+* `dfinsupp.to_finsupp_support`
+
+and therefore maps `finsupp.single` to `dfinsupp.single` and vice versa:
+
+* `finsupp.to_dfinsupp_single`
+* `dfinsupp.to_finsupp_single`
+
+as well as preserving arithmetic operations.
+
+For the bundled equivalences, we provide lemmas that they reduce to `finsupp.to_dfinsupp`:
+
+* `finsupp_add_equiv_dfinsupp_apply`
+* `finsupp_lequiv_dfinsupp_apply`
+* `finsupp_add_equiv_dfinsupp_symm_apply`
+* `finsupp_lequiv_dfinsupp_symm_apply`
+
+## Implementation notes
+
+We provide `dfinsupp.to_finsupp` and `finsupp_equiv_dfinsupp` computably by adding
+`[decidable_eq ι]` and `[Π m : M, decidable (m ≠ 0)]` arguments. To aid with definitional unfolding,
+these arguments are also present on the `noncomputable` equivs.
+-/
+
+variables {ι : Type*} {R : Type*} {M : Type*}
+
+
+/-! ### Basic definitions and lemmas -/
+section defs
+
+/-- Interpret a `finsupp` as a homogenous `dfinsupp`. -/
+def finsupp.to_dfinsupp [has_zero M] (f : ι →₀ M) : Π₀ i : ι, M :=
+⟦⟨f, f.support.1, λ i, (classical.em (f i = 0)).symm.imp_left (finsupp.mem_support_iff.mpr)⟩⟧
+
+@[simp] lemma finsupp.to_dfinsupp_coe [has_zero M] (f : ι →₀ M) : ⇑f.to_dfinsupp = f := rfl
+
+section
+variables [decidable_eq ι] [has_zero M]
+
+@[simp] lemma finsupp.to_dfinsupp_single (i : ι) (m : M) :
+  (finsupp.single i m).to_dfinsupp = dfinsupp.single i m :=
+by { ext, simp [finsupp.single_apply, dfinsupp.single_apply] }
+
+variables [Π m : M, decidable (m ≠ 0)]
+
+@[simp] lemma to_dfinsupp_support (f : ι →₀ M) : f.to_dfinsupp.support = f.support :=
+by { ext, simp, }
+
+/-- Interpret a homogenous `dfinsupp` as a `finsupp`.
+
+Note that the elaborator has a lot of trouble with this definition - it is often necessary to
+write `(dfinsupp.to_finsupp f : ι →₀ M)` instead of `f.to_finsupp`, as for some unknown reason
+using dot notation or ommitting the type ascription prevents the type being resolved correctly. -/
+def dfinsupp.to_finsupp (f : Π₀ i : ι, M) : ι →₀ M :=
+⟨f.support, f, λ i, by simp only [dfinsupp.mem_support_iff]⟩
+
+@[simp] lemma dfinsupp.to_finsupp_coe (f : Π₀ i : ι, M) : ⇑f.to_finsupp = f := rfl
+@[simp] lemma dfinsupp.to_finsupp_support (f : Π₀ i : ι, M) : f.to_finsupp.support = f.support :=
+by { ext, simp, }
+
+@[simp] lemma dfinsupp.to_finsupp_single (i : ι) (m : M) :
+  (dfinsupp.single i m : Π₀ i : ι, M).to_finsupp = finsupp.single i m :=
+by { ext, simp [finsupp.single_apply, dfinsupp.single_apply] }
+
+@[simp] lemma finsupp.to_dfinsupp_to_finsupp (f : ι →₀ M) : f.to_dfinsupp.to_finsupp = f :=
+finsupp.coe_fn_injective rfl
+
+@[simp] lemma dfinsupp.to_finsupp_to_dfinsupp (f : Π₀ i : ι, M) : f.to_finsupp.to_dfinsupp = f :=
+dfinsupp.coe_fn_injective rfl
+
+end
+
+end defs
+
+/-! ### Lemmas about arithmetic operations -/
+section lemmas
+
+namespace finsupp
+
+@[simp] lemma to_dfinsupp_zero [add_zero_class M] :
```

Comment:
Is there a reason you didn't go for just `[has_zero M]`? It still seems to compile.

Reply:
Nice catch.
```suggestion
@[simp] lemma to_dfinsupp_zero [has_zero M] :
```